src/UXClient/Components/HierarchyNavigation/HierarchyNavigation.ts

```diff
@@ -856,12 +855,106 @@ class HierarchyNavigation extends Component{
+    public closeContextMenu = () => {
+        if(this.clickedInstance && this.contextMenu) {
+            this.contextMenu.remove();
             d3.selectAll('li.tsi-selected').classed('tsi-selected', false);
         }
         d3.selectAll('.tsi-modelResultWrapper').classed('tsi-selected', false);
+        this.clickedInstance = null;
+    }
+
+    private prepareForContextMenu = (instanceObj, target, wrapperMousePos, eltMousePos) => {
+        let contextMenuProps = {};
+        contextMenuProps['target'] = target;
+        contextMenuProps['wrapperMousePos'] = wrapperMousePos;
+        contextMenuProps['eltMousePos'] = eltMousePos;
+        this.contextMenuProps = contextMenuProps;
+
+        this.clickedInstance = instanceObj;
+        instanceObj.node.classed('tsi-selected', true);
+    }
+
+    public drawContextMenu = (contextMenuItems, contextMenuOptions: ContextMenuOptions) => {
+        let itemList = [];
+        let contextMenuList;
+        this.contextMenu = this.contextMenuProps['target'].append('div').classed('tsi-hierarchyNavigationContextMenu', true).attr('style', () => `top: ${this.contextMenuProps['wrapperMousePos'] - this.contextMenuProps['eltMousePos']}px`);
+        let renderList = (contextMenuItems) => {
+            if (this.contextMenu.select("ul").empty()) {
+                contextMenuList = this.contextMenu.append('ul');
+            } else {
+                this.contextMenu.select("ul").text('');
+            }
+            
+            contextMenuItems.forEach(item => {
+                let option = item.name;
+                let li = contextMenuList.append('li');
+                
+                if (!contextMenuOptions.isSelectionEnabled) {
+                    li.attr('tabindex', 0)
+                    .attr('arialabel', option)
+                    .attr('title', option)
+                    .text(option).on('click keydown', () => {
+                        if (Utils.isKeyDownAndNotEnter(d3.event)) {return; }
+                        item.action();
+                    });
+                } else {
+                    li.on('click', function() {
+                        let elem = d3.select(this).select(".tsi-hierarchyCheckbox");
+                        if (elem.classed("tsi-notSelected")) {
+                            itemList.push(item);
+                            elem.classed("tsi-notSelected", false);
+                        } else {
+                            let index = itemList.map(elem => elem.name).indexOf(item.name);
+                            itemList.splice(index, 1);
+                            elem.classed("tsi-notSelected", true);
+                        }
+                    })
+                    let div = li.append('div').classed('tsi-selectionItemWrapper', true);
+                    div.append('span').classed('tsi-hierarchyCheckbox tsi-notSelected', true);
+                    div.append('span').classed('tsi-selectionItemKind', true).classed(item.kind, true);
```

Comment:
kinds should go after the variable name, numeric kinds look a bit weird, and aggregate/categorical kinds are kind of busy for a 16x16 space

Reply:
actually i like aggregate and categorical kinds, numeric look pretty bad though right now.  we should also add a title for the kind span explaining what it is

Reply:
waiting for icon updates